### DISCUSSION TITLE
Providing "passwords.yaml" when deploying as a Docker image

### DISCUSSION THREAD

**Discussion by proshin-roman** [serverpod/serverpod#2332] [Q&A]
Hi folks,

I'm trying to deploy my serverpod application to Heroku as a Docker image: my repo contains a slightly modified `Dockerfile` that is then picked up by Heroku, built there and deployed to the cloud. Since the build process is running on the Heroku cloud, it can only access to files committed to the Git repository, hence there is no `passwords.yaml` file in the final Docker image. Then, when the application is started inside the Docker image, it fails to init because of the missing `passwords.yaml` file. 

I wonder if there is a way to replay the `passwords.yaml` file with a set of environment variables - unfortunately, I couldn't find anything in the documentation or in existing Discussions topics.

Am I missing something? Or this feature is simply not supported, and the solution is to somehow push the `passwords.yaml` into the Docker image/container (not secure IMO)? Maybe there is a solution specific to Heroku?

Thank you!

**Comment by proshin-roman** [comment]
Update: I found file `password_manager.dart` in the `serverpod_shared` package that seems to be responsible for loading passwords - its method `Map<String, String>? loadPasswords()` only supports loading from a file... So for me, it looks like there is no other way as to somehow make sure the Docker image contains (has access to) the `passwords.yaml` file.

**Reply by BenAuerDev** [reply]
Hey man, 

I don't know Heroku, but in Github Actions you can use Github-Secrets (base64-encoded) and generate a passwords.yaml for the build process.

Maybe you can create a passwords.yaml file from GitHub Secrets before passing it to the build process? Kind of as a workaround :)

**Reply by Isakdl** [reply]
Hey @proshin-roman, we know about this limitation and are looking into better ways to manage this and provide more options. Such as injecting things as environment variables etc to give more flexibility. You shouldn't keep secrets built into a container anyway, since the image it self then becomes sensitive. 

I haven't worked with Heroku my self, but do they provide a way to mount volumes? If so you could mount the passwords.yaml file in a volume and provide it to the container at the correct path. Then you do not need to build it into the final image.

**Reply by proshin-roman** [reply]
@BenAuerDev certainly that should work, however I don't feel that's the right way: the secrets are then injected into the Docker image that can be - in theory - distributed between different systems. I usually treat Docker images as stateless and secrets-free - it just makes it easier to manage them.

@Isakdl I don't have that much experience with Heroku, so I'm not aware of any out-of-the-fox solution. I've also tried to Google that question - and results also don't look that promising. In the end, using environment variables provides the simplest and the quickest way to configure secrets. It's also given "for free" as having a mountable file system might incur additional costs (well, "it depends" of course). So, looking forward to having that feature built-in into Serverpod 👍
